add example substrate connect page (#1382)
substrate-developer-hub/substrate-docs: content/md/en/docs/reference/substrate-connect.md

---

> title: Substrate Connect

Reviewer: I do not think this is correct. This article is supposed to describe the use/integration of Substrate Connect that integrates Smoldot Light Client. But substrate connect is not a light client per-se, rather an easy way to integrate Smoldot light client into Dapps projects.

> title: Substrate Connect

Reviewer 2: Thanks, @wirednkod! That helps.

> description: Run Wasm Light Clients of any Substrate based chain directly in your browser

Reviewer: Again not really agree with the phrasing. I would rather suggest:
```suggestion
description: Use Substrate Connect in your DApp to interact with any Substrate-based chain.```

> description: Run Wasm Light Clients of any Substrate based chain directly in your browser

Reviewer 2: I would suggest a compromise something like:
Use a light client and add Substrate Connect to your applications to interact with any Substrate-based chain.

keywords:
---
Substrate is a modular framework for creating use-case optimized blockchains at a low cost, by composing custom or pre-built components. Substrate is the backbone that powers Polkadot, a next-generation, heterogeneous, multi-chain network, and its ecosystem.

## Wasm Light Clients

> Substrate-connect turns a browser into a node on a network allowing end-users of Web3 apps to interact with blockchains directly - without connecting to third-party remote nodes and other servers. Removing intermediary servers between network and its users improves security, simplifies infrastructure of a network and lowers its maintenance costs. Decentralized in-browser light clients are a unique feature of substrate-based networks.

Reviewer: Again here the initial description expresses exactly what Substrate connect is, while the suggestion does not. IMO I would add only the key benefits (plus line 17) of the suggestion

[What is a light client and why you should care? by Thibault Sardan](https://www.parity.io/what-is-a-light-client/)

> 

Reviewer: There are 3 things about substrate-connect.
A) Install `@polkadot/rpc-provider` that contains an rpc provider of substrate-connect. With this installation, a user can out-of-the-box use that implementation into migrating from existing use of polkadotJS api to having a light client integrated. (best for Dapps that already use PolkadotJS api, or devs who want to get an app, up and running in no-time)
B) Install `@substrate/connect` this is the actual npm package, that needs severe development in order to work - meaning there is specific handling of sequence of calls that needs to be done (e.g. add relay/wellKnown chain - add parachain in that order etc), handle the JsonRpcCallbacks etc etc (best for adding light client in extensions, desktop apps and/or creating library on top of Substrate/connect)
C) Substrate connect Extension. Extension is actually a "helper" (if I may say), of the dApps that having already integrated substrate connect (Case A above). Extension runs the light client when browser starts, and syncs everything (all well known chains), before a DApp inits. Once a Dapp that contains substrate-connect inits, then Substrate-Connect "understands" the existence of the Extension and uses the already-synced light client of the extension instead of firing up a new one inside the Dapp's tab.

With the above I covered the following topics:
> Install substrate-connect 

Options A and B above depending on the needs of the developer

> Install browser extension

Option C above

> Connect to a well-known blockchain

This is covered in the `content/md/en/docs/reference/substrate-connect.md` file. There are example for either simple usage or advanced usage on how to connect to a well-known chain

> Integrate in apps that use Polkadot-JS API

At the same file as mentioned above, this is covered on paragraph: `Usage with PolkadotJS API (PolkadotJS provider)`

> Add Substrate Connect to other applications

Same answer as above (simple usage, advanced usage)

> ## Well known Networks

Reviewer: Here are mentioned which are the well known networks and it is explained why the id is used and not the name. It is irrelevant to the `Connect to a well-known blockchain` topic

> 

Reviewer: I am not sure I understand this comment

* [Polkadot](https://polkadot.network/) (`polkadot`)
* [Kusama](https://kusama.network/) (`ksmc33`)
* [Westend](https://wiki.polkadot.network/docs/en/maintain-networks#westend-test-network) (`westend2`)
* [Rococo](https://polkadot.network/rococo-v1-a-holiday-gift-to-the-polkadot-community/) (`rococo_v2_2`)

[WellKnownChain enumeration docs](https://paritytech.github.io/substrate-connect/api/enums/connect_src.WellKnownChain.html)

> You have to provide a well-known chain name (`polkadot` for Polkadot, `ksmcc3` for Kusama, `westend2` for Westend and `rococo_v2_2` for Rococo). Note that these are the "real" names of the chains rather than the names they are more commonly known as (such as Kusama or Rococo). For example, "ksmcc3" is the name of Kusama. This is important for chains which have been hard forked. For example, "rococo_v2" and "rococo_v2_2" are two different chains.

Reviewer 2: ```suggestion
Note that you must use the chain name as it appears in the chain code for a specific network rather than the more commonly used network name. 
For example, you must specify `ksmcc3` as the chain name to connect to Kusama. 
Using the correct name is especially important for chains that have been forked.
For example, `rococo_v2` and `rococo_v2_2` are two different chains.
```

Reviewer: Im not sure whI would actually go with:

```suggestion
Note that you must use the chain id as it appears in the chain specification for a specific network rather than the more commonly used network name. 
For example, you must specify `ksmcc3` as the chain identifier to connect to `Kusama`. 
Using the correct name is especially important for chains that have been forked. For example, `rococo_v2` and `rococo_v2_2` are two different chains.
```

> ## Usage with PolkadotJS API (PolkadotJS provider)

Reviewer 2: ```suggestion
## Integrate in apps that use Polkadot-JS API
```

> For using `substrate-connect` with the existing polkadotJS api, there is an rpc-provider that has `substrate-connect` already integrated. You can use that by installing in your app, the following packages:

Reviewer 2: ```suggestion
If you have built applications that use the existing Polkadot-JS API, you can integrate with `substrate-connect` using the `substrate-connect`  RPC provider. 
To add substrate-connect to your application:
```

> #### **@polkadot/rpc-provider**

Reviewer 2: ```suggestion
1. Install the @polkadot/rpc-provider package by running the appropriate command for the package manager you use.
   For example, if you use yarn, run the following command:
```

> This package contains the `substrate-connect` rpc provider that integrates and uses substrate-connect:

Reviewer 2: ```suggestion
```

Reviewer: Im not sure why this should be removed.

```bash
yarn add @polkadot/rpc-provider
```

> *or*

Reviewer 2: Generally it is better to pick a preferred path.

Reviewer: We cannot really - it is the developers decisions if he will use `npm` or `yarn`. We just provide both alternatives. IMO choosing one path would be like, not supporting the alternative (which we do)

> *or*

Reviewer 2: ```suggestion
If you use npm as your package manaager, runn the following command:
```

```bash
npm i @polkadot/rpc-provider
```

> #### **@polkadot/api**

Reviewer 2: ```suggestion
1. Install the `@polkadot/api` package by running the appropriate command for the package manager you use.
   For example, if you use `yarn`, run the following command:
```

> This library provides a clean wrapper around all the methods exposed by a Polkadot/Substrate network client. It is used with the outcome of rpc-provider above

Reviewer 2: ```suggestion
```

```bash
yarn add @polkadot/api
```

> *or*

Reviewer 2: ```suggestion
If you use `npm` as your package manager, run the following command:
```

```bash
npm i @polkadot/api
```

> 

Reviewer 2: ```suggestion
The `polkadot/api` library provides a clean wrapper around all the methods exposed by a Substrate node and used by rpc-provider.

```

Below are some examples that shows how the rpc-provider can be used simply with the already integrated networks (e.g. polkadot, kusama, westend, rococo), or with a custom chain by using its chainspec, or connecting to a parachain."

> *Simple usage (Using an integrated, suported chain)*

Reviewer 2: ```suggestion
### Connect to a well-known chain using Polkadot-JS API
```

Reviewer: That is not correct. User does not connect to a well known chains with polkadot JS API. User connects with Substrate connect provider and then communicates with that chain using the polkadot JS API.

```js
import { ScProvider, WellKnownChain } from '@polkadot/rpc-provider/substrate-connect';
import { ApiPromise } from '@polkadot/api';
// Create the provider for a known chain
const provider = new ScProvider(WellKnownChain.westend2);
// Stablish the connection (and catch possible errors)
await provider.connect()
// Create the PolkadotJS api instance
const api = await ApiPromise.create({ provider });
await api.rpc.chain.subscribeNewHeads((lastHeader) => {
console.log(lastHeader.hash);
});
await api.disconnect();
```

> *Simple usage (custom chain/chainspec)*

Reviewer 2: ```suggestion
### Connect to a custom chain specification using Polkadot-JS API
```

Reviewer: That is not correct. User does not connect to a well known chains with polkadot JS API. User connects with Substrate connect provider and then communicates with that chain using the polkadot JS API.

const provider = new ScProvider(customSpec);
// Stablish the connection (and catch possible errors)
await provider.connect()
// Create the PolkadotJS api instance
const api = await ApiPromise.create({ provider });
await api.rpc.chain.subscribeNewHeads((lastHeader) => {
console.log(lastHeader.hash);
});
await api.disconnect();
```

*Parachains usage*

```js
import { ScProvider, WellKnownChain } from '@polkadot/rpc-provider/substrate-connect';
import { ApiPromise } from '@polkadot/api';
import jsonParachainSpec from './jsonParachainSpec.json';
// Create the provider for the relay chain
const relayProvider = new ScProvider(WellKnownChain.westend2);
// Create the provider for the parachain. Notice that
// we must pass the provider of the relay chain as the
// second argument
const parachainSpec = JSON.stringify(jsonParachainSpec);
const provider = new ScProvider(parachainSpec, relayProvider);
// Stablish the connection (and catch possible errors)
await provider.connect()
// Create the PolkadotJS api instance
const api = await ApiPromise.create({ provider });
await api.rpc.chain.subscribeNewHeads((lastHeader) => {
console.log(lastHeader.hash);
});
await api.disconnect();
```

> ## Advanced Usage

Reviewer 2: ```suggestion
## Add Substrate Connect to other with applications

Reviewer: I think - if we want to change this - it should be something like:
```suggestion
## Use Substrate Connect with other libraries/frameworks
```

> At the section above, `@polkadot/rpc-provider/substrate-connect` was used, because this provider makes if straightforward to create applications that interact with the chain using the Polkadot-JS API. For more advanced application development that doesn't depend on the Polkadot-JS API, you can install and use @substrate-connect. For example, if you are building your own application library or programming interfaces, you should install the Substrate Connect dependencies by running the following command:

Reviewer 2: ```suggestion
The previous section demonstrated how to add the Substrate Connect provider to applications that use the Polkadot-JS API. 
With this provider, you create applications that enable users to interact with the chain through the browser using calls to the Polkadot-JS API methods. 
However, you can install and use @substrate-connect in applications that don't depend on the Polkadot-JS API. For example, if you are building your own application library or programming interfaces, you can install the Substrate Connect dependencies by running the appropriate command for the package manager you use.
```

> 

Reviewer 2: ```suggestion
   For example, if you use yarn, run the following command:

```

```bash
yarn add @substrate/connect
```

> or

Reviewer 2: ```suggestion
If you use `npm` as your package manager, run the following command:
```

```bash
npm i @substrate/connect
```

> *Connecting to a `WellKnownChain`*

Reviewer 2: ```suggestion
### Connect to a well-known chain
```

Reviewer: `WellKnownChain` is a wording/variable that is used inside the codebase and we should probably keep it that way

```suggestion
*Connect to a `WellKnownChain`*
```

Reviewer 2: I'm suggesting we make it a heading and e don't use the back ticks in headings. Even though `WellKnownChain` is a direct reference to code, I don't think it is necessary to call attention to that in a heading. It doesn't improve the readability for someone scanning documentation and it is easier on the human eye to reduce the use of capital letters. When it's not in a heading, though, we can certainly use `WellKnownChain`.

```js
import { WellKnownChain, createScClient } from '@substrate/connect';
// Create the client
const client = createScClient();
// Create the chain connection, while passing the `jsonRpcCallback` function.
const chain = await client.addWellKnownChain(
WellKnownChain.polkadot,
function jsonRpcCallback(response) {
console.log('response', response);
}
);
// send a RpcRequest
chain.sendJsonRpc(
'{"jsonrpc":"2.0","id":"1","method":"system_health","params":[]}'
);
```

> *Connecting to a parachain*

Reviewer 2: ```suggestion
### Connect to a parachain
```
